ledger/ledger-api-client/src/it/scala/com/digitalasset/ledger/client/UpgradesIT.scala

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.client
+
+import com.daml.integrationtest.CantonFixture
+import org.scalatest.Inside
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AsyncWordSpec
+
+import scala.concurrent.{Future}
+import com.google.protobuf.ByteString
+import com.daml.bazeltools.BazelRunfiles._
+import java.io.File
+import java.io.FileInputStream
+import org.scalatest.Suite
+import org.scalatest.compatible.Assertion
+
+import scala.io.Source
+import com.daml.lf.data.Ref.PackageId
+
+import com.daml.lf.archive.{DarReader}
+import scala.util.{Success, Failure}
+
+final class UpgradesIT extends AsyncWordSpec with Matchers with Inside with CantonFixture {
+  self: Suite =>
+  override lazy val devMode = true;
+  override val cantonFixtureDebugMode = true;
+
+  private def loadPackageIdAndBS(path: String): Future[(PackageId, ByteString)] = {
+    val dar = DarReader.assertReadArchiveFromFile(new File(rlocation(path)))
+    assert(dar != null, s"Unable to load test package resource '$path'")
+
+    val testPackage = Future {
+      val in = new FileInputStream(new File(rlocation(path)))
+      assert(in != null, s"Unable to load test package resource '$path'")
+      in
+    }
+    val bytes = testPackage.map(ByteString.readFrom)
+    bytes.onComplete(_ => testPackage.map(_.close()))
+    bytes.map((dar.main.pkgId, _))
+  }
+
+  private def testPackagePair(
+      upgraded: String,
+      upgrading: String,
+      failureMessage: Option[String],
+  ): Future[Assertion] = {
+    for {
+      client <- defaultLedgerClient()
+      (testPackageV1Id, testPackageV1BS) <- loadPackageIdAndBS(upgraded)
+      (testPackageV2Id, testPackageV2BS) <- loadPackageIdAndBS(upgrading)
+      uploadV1Result <- client.packageManagementClient
+        .uploadDarFile(testPackageV1BS)
+        .transform({
+          case Failure(err) => Success(Some(err));
+          case Success(_) => Success(None);
+        })
+      uploadV2Result <- client.packageManagementClient
+        .uploadDarFile(testPackageV2BS)
+        .transform({
+          case Failure(err) => Success(Some(err));
+          case Success(_) => Success(None);
+        })
+    } yield {
+      val cantonLog = Source.fromFile(s"$cantonTmpDir/canton.log").mkString
+      cantonLog should include regex s"Package $testPackageV1Id does not upgrade anything"
+      cantonLog should include regex s"Package $testPackageV2Id upgrades package id $testPackageV1Id"
```

Comment:
These could get quite expensive given how big canton logs can get, is there no way to have the upload response with these statuses?

Reply:
I'm worried about information leakage - if I show a more detailed message than "upgrade failed", someone could upload DARs in order to glean structural information about an uploaded DAR that they shouldn't know about. I assumed only the participant operator should be able to know the reason for upload error.

Reply:
As per the meeting this Wed, information leakage isn't a concern because those with upload privileges should be sufficiently trusted to get that information anyways.

We're getting more than just one id once we get to Carl's changes (checking against the package map) - I'd rather we scan the error message, or the log for this info. I don't expect this to be too expensive - the Canton instances we run in testing are doing relatively few things and not running for very long.